packages/ibm-products/src/components/Checklist/Checklist.js

```diff
@@ -0,0 +1,316 @@
+// or assumption when a prop is not supplied.
+
+// Default values for props
+const defaults = {
+  onClickViewAll: () => {},
+  onToggle: () => {},
+  open: true,
+  showToggle: true,
+  taskLists: [],
+  theme: 'light',
+};
+
+/**
+ * TODO: A description of the component.
+ */
+export let Checklist = React.forwardRef(
+  (
+    {
+      className,
+      onClickViewAll = defaults.onClickViewAll,
+      onToggle = defaults.onToggle,
+      open = defaults.open,
+      chartValue,
+      chartLabel,
+      showToggle = defaults.showToggle,
+      taskLists = defaults.taskLists,
+      theme = defaults.theme,
+      title,
+      viewAllLabel,
+      ...rest
+    },
+    ref
+  ) => {
+    const [isOpen, setIsOpen] = useState(open);
+
+    // Don't use this test: {chartValue && chartLabel && (render...)}.
+    // If `chartValue` is 0 (zero) - a valid value - then it will render 0 and skip the remaining statement.
+    // Use this test instead: {chartLabelAndValue && (render...)}.
+    // The ChecklistChart component will validate `chartValue`.
+    const chartLabelAndValue = typeof chartValue === 'number' && chartLabel;
+
+    const handleClickToggle = () => {
+      setIsOpen((prevOpen) => !prevOpen);
+    };
+
+    const handleClickViewAll = () => {
+      onClickViewAll();
+    };
+
+    // If state changes, then trigger callback.
+    useEffect(() => {
+      onToggle(isOpen);
+    }, [isOpen, onToggle]);
+
+    // If the "open" prop is changed after initialization,
+    // then update internal state.
+    useEffect(() => {
+      setIsOpen(open);
+    }, [open]);
+
+    return (
+      <aside
+        {
+          // Pass through any other property values as HTML attributes.
+          ...rest
+        }
+        className={cx(
+          blockClass, // Apply the block class to the main HTML element
+          className, // Apply any supplied class names to the main HTML element.
+          // example: `${blockClass}__template-string-class-${kind}-n-${size}`,
+          {
+            [`${blockClass}__closed`]: !isOpen,
+          }
+        )}
+        ref={ref}
+        role="main"
+        {...getDevtoolsProps(componentName)}
+      >
+        {(title || chartLabelAndValue) && (
+          <header className={`${blockClass}__header`}>
+            {chartLabelAndValue && (
+              <ChecklistChart value={chartValue} theme={theme} />
+            )}
+
+            <div className={`${blockClass}__titles`}>
+              {title && <h2 className={`${blockClass}__title`}>{title}</h2>}
+
+              {chartLabelAndValue && (
+                <h3 className={`${blockClass}__chart-label`}>{chartLabel}</h3>
+              )}
+            </div>
+
+            {showToggle && (
+              <Button
+                aria-expanded={isOpen}
+                className={`${blockClass}__toggle`}
+                kind="ghost"
+                onClick={handleClickToggle}
+                role="switch"
```

Comment:
Admittedly, `role="switch"` isn't ideal because the lists are not direct descendants of the toggle.

`aria-controls` requires that you refer to another HTML element by its `id`. Using `ids` in component source code is not recommended, since we can't stop a dev from rendering two or more `Checklists` on a page.

Ideas?

Reply:
I think what we sometimes do is create an `id` or `checklistId` prop so that it’s still in the developer’s hands to manage the IDs properly.

Reply:
For the record, after various chats and a meeting: using `uuidv4()` as an internal id generator is ideal.

Was added in [this commit](https://github.com/carbon-design-system/ibm-products/pull/3408/commits/9c897994a9e8ba02db70a0d1ab473b1328977ffc).